# The EARLY BUD-BREAK regulatory module in poplar

# Abstract

Bud-break after winter dormancy is economically and environmentally important process in trees and shrubs from boreal and temperate latitudes but its molecular mechanisms are poorly understood. Previously we have reported the isolation and characterization of the EARLY BUD BREAK 1 (EBB1) gene, encoding an AP2/ERF transcription factor. We show here that EBB1 directly represses the poplar SVL (SHORT VEGETATIVE PHASE-LIKE), a MADS-box protein which was recently found to negatively regulate bud-break in Populus. We also report the identification and characterization of the EARLY BUD-BREAK 3 (EBB3) gene, which encodes another AP2/ERF transcription factor. We demonstrate that EBB3 is a positive regulator of bud-break and acts downstream of EBB1 and SVL. EBB3's effect on bud-break is mediated via control of cell cycle by directly and positively regulating CYCLIND3.1 gene. Our results show the molecular architecture of a novel regulatory module that controls bud-break in poplar by controlling progression of cell cycle.

# Data

The expression data has been deposited at the [ENA](https://ebi.ac.uk/ena) under the accession number **PRJEB35768**

The histone _H3_ and _K27me3_ ChIP-Seq data alignments for the _EBB3_ locus is available in the `data` directory of this repository, including 2 kbp up and downstream of the gene. The gene coordinates themselves were retrieved using [PopGenIE](http://popgenie.org) [1] and the gene ID Potrx003580g02767.

# Quality Assessment

The data was pre-processed following our guidelines [2]. Briefly, the raw data quality was assessed using FastQC [3], reads were filtered for rRNA and trimmed for adapters and quality using SortMeRNA [4] and Trimmomatic [5], respectively. Quantification of expression was then done using kallisto [6]. The quality assessment report generated by MultiQC, summarising the pre-processing (number of reads, alignment rates, _etc._) can be found in the `reports` directory of this repository.

# Code

The scripts used inhouse for the pre-processing are available in the `pipeline` and in the [UPSCb-common](https://github.com/UPSCb/UPSCb-common.git) submodule.

The script used to extract and generate the EBB3 histone profile is available in the `src/R` directory of this repository.

## References
[1] [Sundell _et al._](https://www.ncbi.nlm.nih.gov/pubmed/26192091)
[2] [Delhomme _et al._](https://www.epigenesys.eu/en/protocols/bio-informatics/1283-guidelines-for-rna-seq-data-analysis)
[3] [Andrew _et al._](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
[4] [Kopylova _et al._](https://www.ncbi.nlm.nih.gov/pubmed/23071270)
[5] [Bolger _et al._](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4103590/)
[6] [Bray _et al._](https://www.nature.com/articles/nbt.3519)
